src/tests/ftest/deployment/agent_failure.py

```diff
@@ -0,0 +1,164 @@
+            file_name (str): File name used for self.ior_cmd.test_file.
+        """
+        self.ior_cmd.set_daos_params(self.server_group, self.pool, self.container.uuid)
+        testfile = os.path.join("/", file_name)
+        self.ior_cmd.test_file.update(testfile)
+
+        manager = self.get_ior_job_manager_command()
+        manager.assign_hosts(
+            self.hostlist_clients, self.workdir, self.hostfile_clients_slots)
+
+        try:
+            ior_output = manager.run()
+            results[job_num] = [True]
+            # For debugging.
+            results[job_num].extend(IorCommand.get_ior_metrics(ior_output))
+            # We'll verify the error message.
+            results[job_num].append(ior_output.stderr_text)
+        except CommandFailure as error:
+            results[job_num] = [False, "IOR failed: {}".format(error)]
+
+    def test_agent_failure(self):
+        """Jira ID: DAOS-9385.
+
+        1. Create a pool and a container.
+        2. Run IOR.
+        3. Stop daos_agent process while IOR is running.
+        4. Check the error on the client side. When daos_agent is killed in the middle of
+        IOR, the file write completes successfully, but the pool disconnect at the end
+        fails, so we get the error message that includes -1005. This step is more like a
+        verification of the test itself rather than the product.
+        5. Verify journalctl shows the log that the agent is stopped. Call:
+        journalctl --system -t daos_agent --since <before> --until <after>
+        This step verifies that DAOS, or daos_agent process in this case, prints useful
+        logs for the user to troubleshoot the issue, which in this case the application
+        can’t be used.
+        6. Restart daos_agent.
+        7. Run IOR again. It should succeed this time without any error. This step
+        verifies that DAOS can recover from the fault with minimal human intervention.
+
+        :avocado: tags=all,full_regression
+        :avocado: tags=hw,small
+        :avocado: tags=deployment
+        :avocado: tags=agent_failure
+        """
+        # 1. Create a pool and a container.
+        self.add_pool()
+        self.add_container(self.pool)
+
+        # 2. Run IOR.
+        ior_results = {}
+        job_num = 1
+        self.log.info("Run IOR with thread")
+        job = threading.Thread(
+            target=self.run_ior_collect_error, args=[ior_results, job_num, "test_file_1"])
+
+        self.log.info("Start IOR 1 (thread)")
+        job.start()
+
+        # We need to stop daos_agent while IOR is running, so need to wait for a few
+        # seconds for IOR to start.
+        self.log.info("Sleep 5 sec")
+        time.sleep(5)
+
+        errors = []
+
+        # 3. Stop daos_agent process while IOR is running.
+        since = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+        self.log.info("Stopping agent")
+        stop_agent_errors = self.stop_agents()
```

Comment:
The test should run on multiple server/client environment. And we should have a test case for stopping one agent and verifying what's the behavior and then stopping multiple agents and checking what's the behavior.

Reply:
I think multiple client to check the isolation is a good test case; run two clients with different agent, kill one of them, and verify that it doesn't affect the other. Failure in one client/agent isn't supposed to affect other clients. I created a ticket to cover this case.

I'm not sure stopping multiple agents would add much value, unless I'm missing something.

Reply:
Yeah that's fine. If we cover what you mentioned above.